feat/auto-merge-dev-branch

Keyrxng opened the pull request:
Resolves https://github.com/ubiquity-os/.github/issues/11

- ported over the logic and tests from my original repo
- kept all logic distinctly separate from that of the main plugin to avoid any confusion for future contributors
- the logic is unchanged so the existing QA still stands 
- the only changes I had to make were small ones regarding test config and setup to get the suite green

QA: 

- https://github.com/ubq-testing/auto-merge-dev-branch/actions/runs/19300549187/job/55193555750 [workflow]
- https://github.com/ubq-testing/ubiquibotpav/tree/main [straight merge from dev > main]
- https://github.com/ubq-testing/bot-ai/pull/53 [blocked by conflicts, opens a PR]

Keyrxng commented:
@gentlementlegen, if you could review when you have the time, cheers.

gentlementlegen commented on .github/workflows/auto-merge-development.yml:
```diff
@@ -0,0 +1,35 @@
+    permissions:
+      contents: write
+    steps:
+      - name: Checkout
```
Isn't v5 out yet?

gentlementlegen commented on CI/src/github.ts:
```diff
@@ -0,0 +1,138 @@
+ */
+export async function getDevelopmentBranch(octokit: Octokit, owner: string, repo: string): Promise<BranchData | null> {
+  try {
+    const { data } = await octokit.rest.repos.getBranch({ owner, repo, branch: "development" });
```
Why not fetching the default branch directly?

Keyrxng commented on .github/workflows/auto-merge-development.yml:
```diff
@@ -0,0 +1,35 @@
+          APP_PRIVATE_KEY: ${{ secrets.APP_PRIVATE_KEY }}
+          TARGET_ORGS: ${{ env.TARGET_ORGS }}
+          INACTIVITY_DAYS: ${{ env.INACTIVITY_DAYS }}
+        run: bunx tsx src/cli.ts
```
I forgot we had `bun run`

Keyrxng commented on CI/src/github.ts:
```diff
@@ -0,0 +1,138 @@
+ */
+export async function listOrgRepos(octokit: Octokit, org: string): Promise<RepoData[]> {
+  const repos: RepoData[] = [];
+  for (let page = 1; page <= 10; page++) {
```
using `octokit.paginate` now and updated `msw` handlers to reflect that

Keyrxng commented on CI/src/guards.ts:
```diff
@@ -0,0 +1,43 @@
+      owner: upOwner,
+      repo: upRepo,
+      state: "open",
+      per_page: 1,
```
I figured one was enough to know but I've removed `per_page` now

gentlementlegen commented on CI/src/cli.ts:
```diff
@@ -0,0 +1,38 @@
+import { logger, parseOrgs, requireEnv } from "./utils";
+
+async function main(): Promise<void> {
+  const appId = requireEnv("APP_ID");
```
wouldn't it be nicer to use a typebox schema to read all the value and have the proper types and constraints?

gentlementlegen commented on CI/src/guards.ts:
```diff
@@ -0,0 +1,48 @@
+
+    const parent = repo.parent?.full_name;
+    if (!parent) {
+      logger.info(`[Fork Guard] Fork detected but parent unknown`);
```
all the logs in this function should be `debug` to avoid overly verbose

gentlementlegen commented on CI/src/guards.ts:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    const [upOwner, upRepo] = parent.split("/");
+    const { data: pulls } = await octokit.rest.pulls.list({
```
don't you need to paginate?

gentlementlegen commented on CI/src/main.ts:
```diff
@@ -0,0 +1,354 @@
+  const lastCommitDate = firstValidTimestamp([branch.commit.commit?.committer?.date, branch.commit.commit?.author?.date]);
+
+  if (!lastCommitDate) {
+    logger.info(`[Auto-Merge] Skipping ${repoName}: unable to determine last commit date`);
```
Should be a `warn`

Keyrxng commented:
Thanks for the reviews @gentlementlegen they've been great and I've missed them tbh hahahah. 

Jest CI is failing because of typebox schema/repo env vars but are all green for me locally.

Cheers

gentlementlegen commented:
@Keyrxng For the missing env values you should hard-code placeholders in the action script so that they get populated even in forked pull-requests, and so that the rest of the logic can be tested through mocks if necessary

gentlementlegen commented on CI/src/env.ts:
```diff
@@ -0,0 +1,63 @@
+const ciEnvSchema = Type.Object({
+  APP_ID: Type.String({ minLength: 1 }),
+  APP_PRIVATE_KEY: Type.String({ minLength: 1 }),
+  TARGET_ORGS: Type.Transform(Type.Union([Type.String({ minLength: 1 }), Type.Array(Type.String({ minLength: 1 }), { minItems: 1 })]))
```
Wouldn't always using this an as array make things simpler? The array syntax for one item or many items is the same in Yaml using the dashes, so there is no benefit handling strings

gentlementlegen commented on CI/src/env.ts:
```diff
@@ -0,0 +1,63 @@
+        return input;
+      })
+  ),
+  LOG_LEVEL: Type.Optional(Type.Enum(LOG_LEVEL, { default: LOG_LEVEL.INFO })),
```
No need to set all these as optional for the ones you provide a default value for, using the combination of Default and Decode, they would get their default value set even if none is provided

gentlementlegen commented on .github/workflows/auto-merge-development.yml:
```diff
@@ -0,0 +1,35 @@
+    - cron: "0 6 * * 1" # Mondays at 06:00 UTC
+  workflow_dispatch:
+
+env:
```
Could we read these values from the repo environment and then fallback to these if none are set? My reasonning is that if I want to test it within my org, I have to modify the workflow file instead of just setting the envioronment values

gentlementlegen commented:
Please update the README describing quickly that there is a cron and what it does, which might be useful for users and LLMs to know.

gentlementlegen commented:
@Keyrxng also don't forget to request a review when you're done so I get notifications

gentlementlegen commented on CI/src/types.ts:
```diff
@@ -0,0 +1,35 @@
+
+export type RepositoryInfo = RestEndpointMethodTypes["repos"]["listForOrg"]["response"]["data"][number];
+export type BranchData = RestEndpointMethodTypes["repos"]["getBranch"]["response"]["data"];
+export type Octokit = InstanceType<typeof import("./github").customOctokit>;
```
Why is this needed? Also, it doesn't exist

gentlementlegen commented:
@keyrxng please fix the build issues so I can test it.

gentlementlegen commented on CI/src/summary.ts:
```diff
@@ -0,0 +1,79 @@
+
+  core.summary
+    .addHeading("Auto-Merge Summary")
+    .addRaw(`Processed **${outcomes.length}** repositories with **${errorCount}** errors.`)
```
Nitpick, but you should be probably using `<b>` tags because the `**` are used raw and not translated to bold.
https://github.com/Meniole/daemon-merging/actions/runs/19528379977/attempts/1#summary-55905668588

Keyrxng commented:
@gentlementlegen all resolved now mate

``The development branch has no history in common with main`` -- This error is something I assume we let surface and it can be handled manually if needed. It's too dangerous to merge the histories and force it for production systems imo.

I'm also backing out now if `default === main` as not 100% of repos conform to having `development` as default.

> Seems to work fine. I think on a later version we should not rely on vars in the workflow but rather the configuration, because if an outside org wants to use this functionality, they would have to clone that plugin to set their own TARGET_ORGS

We should also pass in the expected default branch as opposed to fetching it I think, this way we can handle cases better where `default === main`.




QA:

- https://github.com/ubq-testing/automated-merging/actions/runs/19531947690
- https://github.com/ubq-testing/automated-merging/actions/runs/19532735981

<img width="1471" height="685" alt="image" src="https://github.com/user-attachments/assets/0e466c43-f2ca-46eb-8270-a1103a1684f6" />

Keyrxng commented on CI/src/env.ts:
```diff
@@ -0,0 +1,59 @@
+ */
+export function loadConfigEnv(env: Env = process.env) {
+  let orgs: string[] = [];
+  if (env.TARGET_ORGS && typeof env.TARGET_ORGS === "string") {
```
I'm winding down now, will confirm later, but I'm sure Typebox will throw because it's not expecting a `string` and `process.env` will always return a string, so it must be converted either in the schema with `t.transform(t.union(...))` or earlier.

gentlementlegen commented on CI/src/main.ts:
```diff
@@ -0,0 +1,456 @@
+      for (const commit of repoCommits.data) {
+        const authorName = commit.commit.author?.name?.toLowerCase() || "";
+        const committerName = commit.commit.committer?.name?.toLowerCase() || "";
+        if (!authorName.includes("[bot]") && !committerName.includes("[bot]")) {
```
I hope this is safe to assume, would have been better to have the `type` of the user

gentlementlegen commented:
Tried to run it, it seems that all the url the fetch commits were wrong:
https://github.com/Meniole/daemon-merging/actions/runs/19585801894/attempts/1#summary-56094325379

Keyrxng commented:
> Tried to run it, it seems that all the url the fetch commits were wrong:

Fixed, I got my params mixed around after cleaning up the methods, fixed that too so it won't happen again. Also reproducing showed another slight bug with user `user.name` instead of `user.login` causing `isHumanUser` to fail for 0x4007's account.


QA: 

- https://github.com/ubq-testing/automated-merging/actions/runs/19587228102

Hopefully gtm now, ty for the rapid reviews @gentlementlegen. Still feeling rusty but I appreciate you being cool about things.

Keyrxng commented on CI/src/main.ts:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    if (result.aborted) {
+      break;
```
Yeah probably mb [5e9769b](https://github.com/ubiquity-os-marketplace/daemon-merging/pull/60/commits/5e9769b7385e68d6553286311e7a91ebfaead07c)

gentlementlegen commented on CI/src/processing.ts:
```diff
@@ -0,0 +1,214 @@
+  }
+
+  // Skip archived repositories
+  if (repo.archived) {
```
This could be checked before the `forkSafetyGuard` to save on API calls

gentlementlegen reviewed (approved):
Seems good to me.
